strategies/aartur/preparation.py

```diff
@@ -0,0 +1,97 @@
+from rules import State, Wall, Player
+
+
+class AdvancedState:
+    """More convenient State, containing current player and other players."""
+
+    def __init__(self, state: State, player_name: str):
+        self.cells = state.cells
+        self.size_y = len(self.cells)
+        self.size_x = len(self.cells[0])
+
+        self.player: Player
+        self.other_players: list[Player] = []
+        for y in range(self.size_y):
+            for x in range(self.size_x):
+                cell = self.cells[y][x]
+                if isinstance(cell, Player):
+                    if cell.name == player_name:
+                        self.player = cell
+                    else:
+                        self.other_players.append(cell)
+
+    def __getitem__(self, item):
+        return self.cells.__getitem__(item)
+
+
+class MapCell:
+    """Element of a 'Map'.
+
+    Contains distance to the corresponding cell and direction *from* the previous cell
+    (which also contains direction from its previous cell,
+    so one can construct full path from the original cell).
+
+    'visited' flag indicated whether this cell is reachable (False means unreachable)
+    """
+
+    def __init__(self, dist: int = 0, dx: int = 0, dy: int = 0, visited: bool = False):
+        self.dist: int = dist
+        self.dx: int = dx
+        self.dy: int = dy
+        self.visited: bool = visited
+
+
+class Map:
```

Comment:
it looks like we can combine AdvancedState and Map in one class AdvancedState

Reply:
`Map` is used to determine how to move around the board for *one particular player*. I may want to create another `Map` for *another* player to look how he can move around the map. So I will have to duplicate the `AdvancedState` which is bad.

Reply:
ExtendedState must contain ReachabilityGraph, i think

Reply:
Renamed to `ReachabilityGraph` and `ExtendedState`.
I don't think that State should contain Graph: they are used for different purposes and one may need Graphs for all players while other might don't need Graph at all.